src/rust/lib_ccxr/src/util/log.rs

```diff
@@ -0,0 +1,592 @@
+            ExitCause::Unsupported => 9,
+            ExitCause::NotEnoughMemory => 500,
+            ExitCause::Bug => 1000,
+        }
+    }
+}
+
+impl<'a> CCExtractorLogger {
+    /// Returns a new instance of CCExtractorLogger with the provided settings.
+    ///
+    /// `gui_mode` is used to determine if the log massages are intercepted by a GUI.
+    /// `target` specifies the location for printing the log messages.
+    /// `debug_mask` is used to filter debug messages based on its type.
+    pub const fn new(
+        target: OutputTarget,
+        debug_mask: DebugMessageMask,
+        gui_mode: bool,
+    ) -> CCExtractorLogger {
+        CCExtractorLogger {
+            target,
+            debug_mask,
+            gui_mode,
+        }
+    }
+
+    /// Set the mode to Normal or Debug Mode based on `false` or `true` respectively for the
+    /// underlying [`DebugMessageMask`].
+    ///
+    /// This method switches the mask used for filtering debug messages.
+    /// Similar to [`DebugMessageMask::set_debug_mode`].
+    pub fn set_debug_mode(&mut self, mode: bool) {
+        self.debug_mask.set_debug_mode(mode)
+    }
+
+    /// Check if the current mode is set to Debug Mode.
+    ///
+    /// Similar to [`DebugMessageMask::is_debug_mode`].
+    pub fn is_debug_mode(&self) -> bool {
+        self.debug_mask.is_debug_mode()
+    }
+
+    /// Returns the currently set target for logging messages.
+    pub fn target(&self) -> OutputTarget {
+        self.target
+    }
+
+    /// Check if the messages are intercepted by GUI.
+    pub fn is_gui_mode(&self) -> bool {
+        self.gui_mode
+    }
+
+    fn print(&self, args: &Arguments<'a>) {
+        match &self.target {
+            OutputTarget::Stdout => print!("{}", args),
+            OutputTarget::Stderr => eprint!("{}", args),
+            OutputTarget::Quiet => {}
+        }
+    }
+
+    /// Log a fatal error message. Use [`fatal!`] instead.
+    ///
+    /// Used for logging errors dangerous enough to crash the program instantly.
+    pub fn log_fatal(&self, exit_cause: ExitCause, args: &Arguments<'a>) -> ! {
+        self.log_error(args);
+        info!("Issues? Open a ticket here");
+        info!("https://github.com/CCExtractor/ccextractor/issues");
+        std::process::exit(exit_cause.exit_code())
+    }
+
+    /// Log an error message. Use [`error!`] instead.
+    ///
+    /// Used for logging general errors occuring in the program.
+    pub fn log_error(&self, args: &Arguments<'a>) {
+        if self.gui_mode {
+            eprint!("###MESSAGE#")
+        } else {
+            eprint!("\rError: ")
+        }
+
+        eprintln!("{}", args);
+    }
+
+    /// Log an informational message. Use [`info!`] instead.
+    ///
+    /// Used for logging extra information about the execution of the program.
+    pub fn log_info(&self, args: &Arguments<'a>) {
+        if self.target == OutputTarget::Quiet {
+            return;
+        }
+
+        activity_header();
```

Comment:
I don't think we need to print the `activity_header()` here anymore. We can just print it once at the start, and remove it from being called for each `info!` log

Reply:
did it